safekeeper/src/wal_reader_stream.rs

```diff
@@ -0,0 +1,145 @@
+use std::sync::Arc;
+
+use async_stream::try_stream;
+use bytes::Bytes;
+use futures::Stream;
+use postgres_backend::CopyStreamHandlerEnd;
+use std::time::Duration;
+use tokio::time::timeout;
+use utils::lsn::Lsn;
+
+use crate::{
+    safekeeper::Term,
+    send_wal::{EndWatch, WalSenderGuard},
+    timeline::WalResidentTimeline,
+};
+
+pub(crate) struct WalReaderStreamBuilder {
+    pub(crate) tli: WalResidentTimeline,
+    pub(crate) start_pos: Lsn,
+    pub(crate) end_pos: Lsn,
+    pub(crate) term: Option<Term>,
+    pub(crate) end_watch: EndWatch,
+    pub(crate) wal_sender_guard: Arc<WalSenderGuard>,
+}
+
+impl WalReaderStreamBuilder {
+    pub(crate) fn start_pos(&self) -> Lsn {
+        self.start_pos
+    }
+}
+
+pub(crate) struct WalBytes {
+    /// Raw PG WAL
+    pub(crate) wal: Bytes,
+    /// Start LSN of [`Self::wal`]
+    pub(crate) wal_start_lsn: Lsn,
+    /// End LSN of [`Self::wal`]
+    pub(crate) wal_end_lsn: Lsn,
+    /// End LSN of WAL available on the safekeeper
+    pub(crate) commit_lsn: Lsn,
```

Comment:
This is not ideal naming because while pageserver fetched WAL only up to commit_lsn, compute does it to flush_lsn -- so it it would be wrong once send_wal.rs also starts using wal_reader_stream. I'd probably drop this field completely since InterpretedWalSender has end_watch_view.

The note is partially applies to the commit_lsn name in InterpretedWalRecordsBody; similarly, since only pageserver uses this proto it is indeed technically commit_lsn, but generally WAL serving can give more then that. So I'd either leave old `wal_end` name (which is also used in postgres protocol), or add a comment explaining this.

Reply:
Good point. Renamed in https://github.com/neondatabase/neon/pull/9746/commits/e659f81e2c899820b10e84a8b5cf10e4a7ea9011. We can refactor separately to avoid the wal reader producing `commit_lsn`.